From a35ff85301444e068a80bae7698ad1bd9e81788b Mon Sep 17 00:00:00 2001
From: Jianing Ren <jianing.ren@rock-chips.com>
Date: Fri, 5 Jun 2020 18:55:11 +0800
Subject: [PATCH] phy: phy-rockchip-naneng-usb2: enable id irq when vbus always
 on

This patch will enable id interrupt for otg port when the property
vbus-always-on is set in dts.

If vbus is always on, bvalid interrupt won't make sense and be
disabled. But we cat change id state to switch drd mode by software.

Change-Id: I69fc3f00430dfd73835c6e99694d8d7c72c60c8c
Signed-off-by: Jianing Ren <jianing.ren@rock-chips.com>
---
 .../phy/rockchip/phy-rockchip-naneng-usb2.c   | 67 +++++++++----------
 1 file changed, 33 insertions(+), 34 deletions(-)

diff --git a/drivers/phy/rockchip/phy-rockchip-naneng-usb2.c b/drivers/phy/rockchip/phy-rockchip-naneng-usb2.c
index d2c3ce9fe6da..04cb59b6274e 100644
--- a/drivers/phy/rockchip/phy-rockchip-naneng-usb2.c
+++ b/drivers/phy/rockchip/phy-rockchip-naneng-usb2.c
@@ -1052,6 +1052,39 @@ static int rockchip_usb2phy_otg_port_init(struct rockchip_usb2phy *rphy,
 		goto out;
 	}
 
+	/* Request otg iddig interrupt only if there is no extcon property */
+	if (rphy->edev_self) {
+		rport->id_irq = of_irq_get_byname(child_np, "otg-id");
+		if (rport->id_irq <= 0) {
+			dev_err(rphy->dev, "no otg id irq provided\n");
+			return -EINVAL;
+		}
+
+		ret = devm_request_threaded_irq(rphy->dev,
+						rport->id_irq, NULL,
+						rockchip_usb2phy_id_irq,
+						IRQF_ONESHOT,
+						"rockchip_usb2phy_id",
+						rport);
+		if (ret) {
+			dev_err(rphy->dev,
+				"failed to request otg-id irq handle\n");
+			return ret;
+		}
+
+		iddig = property_enabled(rphy->grf,
+					 &rport->port_cfg->utmi_iddig);
+		if (!iddig) {
+			extcon_set_state(rphy->edev, EXTCON_USB, false);
+			extcon_set_state(rphy->edev, EXTCON_USB_HOST, true);
+			extcon_set_state(rphy->edev, EXTCON_USB_VBUS_EN, true);
+			/* Enable VBUS supply */
+			ret = rockchip_set_vbus_power(rport, true);
+			if (ret)
+				return ret;
+		}
+	}
+
 	if (rport->vbus_always_on)
 		goto out;
 
@@ -1076,40 +1109,6 @@ static int rockchip_usb2phy_otg_port_init(struct rockchip_usb2phy *rphy,
 
 	INIT_DELAYED_WORK(&rport->otg_sm_work, rockchip_usb2phy_otg_sm_work);
 
-	if (!rphy->edev_self)
-		goto out;
-
-	/* Request otg iddig interrupt only if there is no extcon property */
-	rport->id_irq = of_irq_get_byname(child_np, "otg-id");
-	if (rport->id_irq <= 0) {
-		dev_err(rphy->dev, "no otg id irq provided\n");
-		return -EINVAL;
-	}
-
-	ret = devm_request_threaded_irq(rphy->dev,
-					rport->id_irq, NULL,
-					rockchip_usb2phy_id_irq,
-					IRQF_ONESHOT,
-					"rockchip_usb2phy_id",
-					rport);
-	if (ret) {
-		dev_err(rphy->dev,
-			"failed to request otg-id irq handle\n");
-		return ret;
-	}
-
-	iddig = property_enabled(rphy->grf,
-				 &rport->port_cfg->utmi_iddig);
-	if (!iddig) {
-		extcon_set_state(rphy->edev, EXTCON_USB, false);
-		extcon_set_state(rphy->edev, EXTCON_USB_HOST, true);
-		extcon_set_state(rphy->edev, EXTCON_USB_VBUS_EN, true);
-		/* Enable VBUS supply */
-		ret = rockchip_set_vbus_power(rport, true);
-		if (ret)
-			return ret;
-	}
-
 out:
 	/*
 	 * Let us put phy-port into suspend mode here for saving power
-- 
2.35.3

